tests/test_application.py

```diff
@@ -0,0 +1,1694 @@
+#!/usr/bin/env python
+#
+# A library that provides a Python interface to the Telegram Bot API
+# Copyright (C) 2015-2022
+# Leandro Toledo de Souza <devs@python-telegram-bot.org>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser Public License for more details.
+#
+# You should have received a copy of the GNU Lesser Public License
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+"""The integration of persistence into the application is tested in test_basepersistence.
+"""
+import asyncio
+import inspect
+import logging
+import os
+import platform
+import signal
+import threading
```

Comment:
why are we using threading in many of the tests here?

Reply:
that's for testing `App.run_*`. because those call `loop.run_forever`, they block the main thread and to make assertions about the app running, I found no other solution than to go to a different thread …